Modules/posixmodule.c

```diff
@@ -5088,384 +5088,358 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
+static BOOL
+_testInfo(DWORD attributes, DWORD reparseTag, int testedType)
+{
+    if (testedType == PY_IFREG) {
+        return attributes && !(attributes & FILE_ATTRIBUTE_DIRECTORY);
+    }
+    if (testedType == PY_IFDIR) {
+        return attributes & FILE_ATTRIBUTE_DIRECTORY;
+    }
+    if (attributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+        switch (testedType) {
+        case PY_IFLNK:
+            return reparseTag == IO_REPARSE_TAG_SYMLINK;
+        case PY_IFMNT:
+            return reparseTag == IO_REPARSE_TAG_MOUNT_POINT;
+        case PY_IFLRP:
+            return IsReparseTagNameSurrogate(reparseTag);
+        case PY_IFRRP:
+            return reparseTag && !IsReparseTagNameSurrogate(reparseTag);
+        }
+    }
 
-[clinic start generated code]*/
+    return FALSE;
+}
 
-static PyObject *
-os__path_isdir_impl(PyObject *module, PyObject *s)
-/*[clinic end generated code: output=9d87ab3c8b8a4e61 input=c17f7ef21d22d64e]*/
+static BOOL
+_testFileTypeByHandle(HANDLE hfile, int testedType, BOOL diskOnly)
 {
-    HANDLE hfile;
-    BOOL close_file = TRUE;
+    assert(testedType == PY_IFREG || testedType == PY_IFDIR ||
+           testedType == PY_IFLNK || testedType == PY_IFMNT ||
+           testedType == PY_IFLRP || testedType == PY_IFRRP);
+
+    if (diskOnly && GetFileType(hfile) != FILE_TYPE_DISK) {
+        return FALSE;
+    }
+
+    if (testedType != PY_IFREG && testedType != PY_IFDIR) {
+        FILE_ATTRIBUTE_TAG_INFO info;
+        return GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
+                                            sizeof(info)) &&
+               _testInfo(info.FileAttributes, info.ReparseTag, testedType);
+    }
     FILE_BASIC_INFO info;
-    path_t _path = PATH_T_INITIALIZE("isdir", "s", 0, 1);
-    int result;
-    BOOL slow_path = TRUE;
-    FILE_STAT_BASIC_INFORMATION statInfo;
+    return GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
+                                        sizeof(info)) &&
+           _testInfo(info.FileAttributes, 0, testedType);
+}
 
-    if (!path_converter(s, &_path)) {
-        path_cleanup(&_path);
-        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
-            PyErr_Clear();
-            Py_RETURN_FALSE;
+static BOOL
+_testFileTypeByName(LPCWSTR path, int testedType)
+{
+    assert(testedType == PY_IFREG || testedType == PY_IFDIR ||
+           testedType == PY_IFLNK || testedType == PY_IFMNT ||
+           testedType == PY_IFLRP || testedType == PY_IFRRP);
+
+    FILE_STAT_BASIC_INFORMATION info;
+    if (_Py_GetFileInformationByName(path, FileStatBasicByNameInfo, &info,
+                                     sizeof(info)))
+    {
+        BOOL result = _testInfo(info.FileAttributes, info.ReparseTag,
+                                testedType);
+        if (!result || (testedType != PY_IFREG && testedType != PY_IFDIR) ||
+            !(info.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT))
+        {
+            return result;
         }
```

Comment:
I overlooked a false positive for `PY_IFREG` if it's not a file in a disk filesystem, such as a named pipe (e.g. "\\\\.\\PIPE\\InitShutdown") or a mailslot (e.g. "\\\\.\\MAILSLOT\\WAITFOR.EXE\\StartSignal"). This will be an issue if NTAPI `NtQueryInformationByName()` and `FileStatBasicByNameInfo` ever get implemented by the named-pipe filesystem or mailslot filesystem, or by any other filesystem with 'files' that shouldn't be reported as regular files. It's already an issue for `_testFileTypeByHandle()`.

The low-level NT `info.DeviceType` should be checked for a disk device, i.e. `diskDevice = (info.DeviceType == FILE_DEVICE_DISK || info.DeviceType == FILE_DEVICE_VIRTUAL_DISK || info.DeviceType == FILE_DEVICE_CD_ROM)`. Also, `_testFileTypeByHandle()` will need to always set `diskDevice`, i.e. `diskDevice = (GetFileType(hfile) == FILE_TYPE_DISK)`. Add the boolean `diskDevice` parameter to `_testInfo()` to gate the `PY_IFREG` test, i.e. `diskDevice && attributes && !(attributes & FILE_ATTRIBUTE_DIRECTORY)`.